File: Collectr/src/main/java/io/jahidem/collectr/model/Collection.java
package io.jahidem.collectr.model;

import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.FullTextField;
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.IndexedEmbedded;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;


@Entity
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
@JsonIgnoreProperties(value = {"hibernateLazyInitializer", "handler"})
public class Collection {

    @JsonIgnore
    @OneToMany(mappedBy = "collection",
            cascade = CascadeType.ALL,
            fetch = FetchType.LAZY)
    List<Item> items;
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;
    @FullTextField(analyzer = "english")
    private String title;
    @FullTextField(analyzer = "english")
    private String description;
    private String imageId;
    @OneToOne(
            mappedBy = "collection",
            fetch = FetchType.LAZY,
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    private ItemTemplate itemTemplate;

    @IndexedEmbedded(includeDepth = 1)
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(
            name = "catagory_id"
    )
    private CollectionCatagory catagory;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;
    @CreationTimestamp
    @Column(
            updatable = false
    )
    private LocalDateTime createdAt;

}

File: frontend/src/types/collection.ts
import { User } from "./auth";



export type Collection = {
  id: string;
  title: string;
  description: string;
  imageId: string;
  catagory: CollectionCatagory;
  itemTemplate: ItemTemplate;
  user: User;
}


export type CollectionCatagory = {
  id: string
  name: string;
}

export type ItemTemplate = {
  id: string,
  collection: string;
  itemFields: ItemField[];

}

export type ItemField = {
  id: string;
  fieldName: string;
  itemFieldType: ItemFieldType;
  fieldValue: string | null;
}

export enum ItemFieldType {
  SELECT_FIELD = "SELECT_FIELD",
  INTEGER_FIELD = "INTEGER_FIELD",
  STRING_FIELD = "STRING_FIELD",
  MULTILINE_STRING_FIELD = "MULTILINE_STRING_FIELD",
  BOOLEAN_FIELD = "BOOLEAN_FIELD",
  DATE_FIELD = "DATE_FIELD"
}